crates/kitsune_p2p/proxy/src/config.rs

```diff
@@ -0,0 +1,156 @@
+use crate::*;
+
+/// Callback function signature for proxy accept/deny.
+pub type AcceptProxyCallbackFn =
+    Arc<dyn Fn(CertDigest) -> MustBoxFuture<'static, bool> + 'static + Send + Sync>;
+
+/// Callback type for proxy accept/deny.
+#[derive(Clone, Deref, AsRef)]
+pub struct AcceptProxyCallback(pub AcceptProxyCallbackFn);
+
+impl AcceptProxyCallback {
+    /// Callback that blanket denies all proxy requests.
+    pub fn reject_all() -> Self {
+        Self(Arc::new(|_| async { false }.boxed().into()))
+    }
+
+    /// Callback that blanket accepts all proxy requests.
+    pub fn accept_all() -> Self {
+        Self(Arc::new(|_| async { true }.boxed().into()))
+    }
+}
+
+/// Tls Configuration for proxy.
+#[derive(Clone)]
+pub struct TlsConfig {
+    /// Cert
+    pub cert: Cert,
+
+    /// Cert Priv Key
+    pub cert_priv_key: CertPrivKey,
+
+    /// Cert Digest
+    pub cert_digest: CertDigest,
+}
+
+impl TlsConfig {
+    /// Create a new ephemeral tls certificate that will not be persisted.
+    pub async fn new_ephemeral() -> TransportResult<Self> {
+        let mut options = lair_keystore_api::actor::TlsCertOptions::default();
+        options.alg = lair_keystore_api::actor::TlsCertAlg::PkcsEcdsaP256Sha256;
+        let cert = lair_keystore_api::internal::tls::tls_cert_self_signed_new_from_entropy(options)
+            .await
+            .map_err(TransportError::other)?;
+        Ok(Self {
+            cert: cert.cert_der,
+            cert_priv_key: cert.priv_key_der,
+            cert_digest: cert.cert_digest,
+        })
+    }
+}
+
+/// Configuration for proxy binding.
+pub enum ProxyConfig {
+    /// We want to be hosted at a remote proxy location.
+    RemoteProxyClient {
+        /// The Tls config for this proxy endpoint.
+        tls: TlsConfig,
+
+        /// The remote proxy url to be hosted at.
+        proxy_url: ProxyUrl,
+    },
+
+    /// We want to be a proxy server for others.
+    /// (We can also deny all proxy requests for something in-between).
+    LocalProxyServer {
```

Comment:
This means I can be directly called (like 192.168.0.4:9000) on a LAN for example? What is an example of this mode acting as a proxy for someone else?
Or could this be on a public ip?

Reply:
Yeah, this can work with LANs or if the machine you are on has a directly addressable IP.

We will need a way of passing in an override for NAT cases where the router/firewall is forwarding a different IP to your LAN address... or if you want a vanity domain name instead of a raw IP address : )

But that will be another story.